<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genophenocorr.preprocessing._variant &mdash; genophenocorr 0.1.0 documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/genophenocorr.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/jquery.js?v=5d32c60e"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../../../_static/documentation_options.js?v=01f34227"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../../../_static/copybutton.js?v=6dbb43f8"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            genophenocorr
          </a>
              <div class="version">
                0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Installation &amp; Tutorial</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../user-guide/index.html">User guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../apidocs/modules.html">API reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">genophenocorr</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">genophenocorr.preprocessing._variant</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for genophenocorr.preprocessing._variant</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="kn">import</span> <span class="nn">typing</span>

<span class="kn">import</span> <span class="nn">hpotk</span>
<span class="kn">import</span> <span class="nn">requests</span>


<span class="kn">from</span> <span class="nn">genophenocorr.model</span> <span class="kn">import</span> <span class="n">VariantCoordinates</span><span class="p">,</span> <span class="n">TranscriptAnnotation</span><span class="p">,</span> <span class="n">Variant</span>

<span class="kn">from</span> <span class="nn">._api</span> <span class="kn">import</span> <span class="n">FunctionalAnnotator</span><span class="p">,</span> <span class="n">ProteinMetadataService</span>


<span class="k">def</span> <span class="nf">verify_start_end_coordinates</span><span class="p">(</span><span class="n">vc</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Converts the 0-based VariantCoordinates to ones that will be interpreted correctly by VEP</span>
<span class="sd">    </span>
<span class="sd">    Example - an insertion/duplication of G after the given G at coordinate 3:</span>
<span class="sd">    1 2 3 4 5</span>
<span class="sd">    A C G T A</span>

<span class="sd">    0-based: 2 3 G GG       1-based: 3 G GG         VEP: 4 3 - G </span>

<span class="sd">    Args:</span>
<span class="sd">        vc (VariantCoordinates): A VariantCoordinates object</span>
<span class="sd">    Returns:</span>
<span class="sd">        string: The variant coordinates formatted to work with VEP</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">chrom</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">chrom</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">end</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">start</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="n">alt</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">alt</span>
    <span class="k">if</span> <span class="n">vc</span><span class="o">.</span><span class="n">is_structural</span><span class="p">():</span>
        <span class="n">alt</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">alt</span><span class="p">[</span><span class="mi">1</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="c1"># TODO: Verify &lt;INS&gt; are working correctly</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vc</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">vc</span><span class="o">.</span><span class="n">alt</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Trimmed alleles are not yet supported!&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">vc</span><span class="o">.</span><span class="n">ref</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">vc</span><span class="o">.</span><span class="n">alt</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># INS/DUP</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">1</span>  <span class="c1"># we must &quot;trim&quot;</span>
            <span class="n">end</span> <span class="o">=</span> <span class="n">end</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">alt</span> <span class="o">=</span> <span class="n">vc</span><span class="o">.</span><span class="n">alt</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span>
            <span class="c1"># 100 AC AGT</span>
            <span class="c1"># MNV</span>

    <span class="k">return</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">chrom</span><span class="si">}</span><span class="s1">:</span><span class="si">{</span><span class="n">start</span><span class="si">}</span><span class="s1">-</span><span class="si">{</span><span class="n">end</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">alt</span><span class="si">}</span><span class="s1">&#39;</span>


<div class="viewcode-block" id="VepFunctionalAnnotator">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VepFunctionalAnnotator">[docs]</a>
<span class="k">class</span> <span class="nc">VepFunctionalAnnotator</span><span class="p">(</span><span class="n">FunctionalAnnotator</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A class that creates a Variant object using variant coordinates and Variant Effect Predictor (VEP) REST API</span>

<span class="sd">    Methods:</span>
<span class="sd">        annotate(variant_coordinates: VariantCoordinates): Creates a Variant object from VariantCoordinates</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">protein_annotator</span><span class="p">:</span> <span class="n">ProteinMetadataService</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructs all necessary attributes for a VepFunctionalAnnotator object</span>

<span class="sd">        Args:</span>
<span class="sd">            protein_annotator (ProteinMetadataService): A ProteinMetadataService object for ProteinMetadata creation</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_logging</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_protein_annotator</span> <span class="o">=</span> <span class="n">protein_annotator</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_url</span> <span class="o">=</span> <span class="s1">&#39;https://rest.ensembl.org/vep/human/region/</span><span class="si">%s</span><span class="s1">?LoF=1&amp;canonical=1&amp;domains=1&amp;hgvs=1&#39;</span> \
                    <span class="s1">&#39;&amp;mutfunc=1&amp;numbers=1&amp;protein=1&amp;refseq=1&amp;transcript_version=1&amp;variant_class=1&#39;</span>

<div class="viewcode-block" id="VepFunctionalAnnotator.annotate">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VepFunctionalAnnotator.annotate">[docs]</a>
    <span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Variant</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a Variant object by searching variant coordinates with Variant Effect Predictor (VEP) REST API. </span>

<span class="sd">        Args:</span>
<span class="sd">            variant_coordinates (VariantCoordinates): A VariantCoordinates object</span>
<span class="sd">        Returns:</span>
<span class="sd">            Variant: A Variant object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variant</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_query_vep</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">)</span>
        <span class="n">variant_id</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;id&#39;</span><span class="p">)</span>
        <span class="n">variant_class</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;variant_class&#39;</span><span class="p">)</span>
        <span class="n">canon_tx</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">transcript_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">trans</span> <span class="ow">in</span> <span class="n">variant</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;transcript_consequences&#39;</span><span class="p">):</span>
            <span class="n">trans_id</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;transcript_id&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">trans_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;NM&#39;</span><span class="p">):</span>
                <span class="k">continue</span>
            <span class="k">if</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;canonical&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="n">canon_tx</span> <span class="o">=</span> <span class="n">trans_id</span>
            <span class="n">hgvsc_id</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;hgvsc&#39;</span><span class="p">)</span>
            <span class="n">consequences</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;consequence_terms&#39;</span><span class="p">)</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">)</span>
            <span class="n">protein_id</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;protein_id&#39;</span><span class="p">)</span>
            <span class="n">protein</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_protein_annotator</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
            <span class="n">protein_effect_start</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;protein_start&#39;</span><span class="p">)</span>
            <span class="n">protein_effect_end</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;protein_end&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">protein_effect_start</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">protein_effect_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">protein_effect_start</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">if</span> <span class="n">protein_effect_end</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">protein_effect_end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">protein_effect_end</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">protein_effect_start</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">protein_effect_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">protein_effect_start</span><span class="p">)</span>
            <span class="n">exons_effected</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;exon&#39;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">exons_effected</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">exons_effected</span> <span class="o">=</span> <span class="n">exons_effected</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;-&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">exons_effected</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
                    <span class="n">exons_effected</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">exons_effected</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="nb">int</span><span class="p">(</span><span class="n">exons_effected</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
                <span class="n">exons_effected</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">exons_effected</span><span class="p">]</span>
            <span class="n">transcript_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">TranscriptAnnotation</span><span class="p">(</span><span class="n">gene_name</span><span class="p">,</span>
                                     <span class="n">trans_id</span><span class="p">,</span>
                                     <span class="n">hgvsc_id</span><span class="p">,</span>
                                     <span class="n">consequences</span><span class="p">,</span>
                                     <span class="n">exons_effected</span><span class="p">,</span>
                                     <span class="n">protein</span><span class="p">,</span>
                                     <span class="n">protein_effect_start</span><span class="p">,</span>
                                     <span class="n">protein_effect_end</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">Variant</span><span class="p">(</span><span class="n">variant_id</span><span class="p">,</span> <span class="n">variant_class</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">,</span> <span class="n">transcript_list</span><span class="p">,</span>
                       <span class="n">variant_coordinates</span><span class="o">.</span><span class="n">genotype</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_query_vep</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
        <span class="n">api_url</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_url</span> <span class="o">%</span> <span class="p">(</span><span class="n">verify_start_end_coordinates</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">))</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">api_url</span><span class="p">,</span> <span class="n">headers</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;Content-Type&#39;</span><span class="p">:</span> <span class="s1">&#39;application/json&#39;</span><span class="p">})</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">r</span><span class="o">.</span><span class="n">ok</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected a result but got an Error for variant: </span><span class="si">{</span><span class="n">variant_coordinates</span><span class="o">.</span><span class="n">as_string</span><span class="p">()</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">r</span><span class="o">.</span><span class="n">raise_for_status</span><span class="p">()</span>
        <span class="n">results</span> <span class="o">=</span> <span class="n">r</span><span class="o">.</span><span class="n">json</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">results</span><span class="p">,</span> <span class="nb">list</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">results</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s1">&#39;error&#39;</span><span class="p">))</span>
            <span class="k">raise</span> <span class="ne">ConnectionError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected a result but got an Error. See log for details.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_logging</span><span class="o">.</span><span class="n">error</span><span class="p">([</span><span class="n">result</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="p">])</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Expected only one variant per request but received </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">results</span><span class="p">)</span><span class="si">}</span><span class="s2"> different variants.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">results</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span></div>



<div class="viewcode-block" id="VariantAnnotationCache">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VariantAnnotationCache">[docs]</a>
<span class="k">class</span> <span class="nc">VariantAnnotationCache</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A class that stores or retrieves Variant objects using pickle format</span>

<span class="sd">    Methods:</span>
<span class="sd">        get_annotations(variant_coordinates:VariantCoordinates): Searches a given data directory for a pickle file with variant coordinates and returns a Variant object</span>
<span class="sd">        store_annotations(variant_coordinates:VariantCoordinates, annotation:Variant): Creates a pickle file with variant coordinates and stores the given Variant object into that file  </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">datadir</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructs all necessary attributes for a VariantAnnotationCache object</span>

<span class="sd">        Args:</span>
<span class="sd">            datadir (string): A string that references an existing directory that does or will contain all pickle files being stored</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">datadir</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;datadir </span><span class="si">{</span><span class="n">datadir</span><span class="si">}</span><span class="s1"> must be an existing directory&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_datadir</span> <span class="o">=</span> <span class="n">datadir</span>

<div class="viewcode-block" id="VariantAnnotationCache.get_annotations">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VariantAnnotationCache.get_annotations">[docs]</a>
    <span class="k">def</span> <span class="nf">get_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">typing</span><span class="o">.</span><span class="n">Optional</span><span class="p">[</span><span class="n">Variant</span><span class="p">]:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Searches a given data directory for a pickle file with given variant coordinates and returns Variant from file. Returns None if no file is found.</span>

<span class="sd">        Args:</span>
<span class="sd">            variant_coordinates (VariantCoordinates): The variant_coordinates associated with the desired Variant</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fpath</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_file_name</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fpath</span><span class="p">):</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fpath</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
                <span class="k">return</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">fh</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span></div>


<div class="viewcode-block" id="VariantAnnotationCache.store_annotations">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VariantAnnotationCache.store_annotations">[docs]</a>
    <span class="k">def</span> <span class="nf">store_annotations</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">,</span> <span class="n">annotation</span><span class="p">:</span> <span class="n">Variant</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a pickle file with the given variant coordinates in the file name. Loads the Variant object given into the file for storage.</span>

<span class="sd">        Args:</span>
<span class="sd">            variant_coordinates (VariantCoordinates): The variant_coordinates associated with the desired Variant</span>
<span class="sd">            annotation (Variant): A Variant object that will be stored under the given variant coordinates</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fpath</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_create_file_name</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">fpath</span><span class="p">,</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="n">annotation</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span></div>


    <span class="k">def</span> <span class="nf">_create_file_name</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a file name with full location and the variant coordinates (e.g. &quot;/path/to/desired/directory/1_2345_G_C_heterozygous.pickle&quot;)</span>

<span class="sd">        Args:</span>
<span class="sd">            variant_coordinates (VariantCoordinates): The variant coordinates associated with the Variant</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fname</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">variant_coordinates</span><span class="o">.</span><span class="n">as_string</span><span class="p">()</span><span class="si">}</span><span class="s1">.pickle&#39;</span>
        <span class="k">return</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_datadir</span><span class="p">,</span> <span class="n">fname</span><span class="p">)</span></div>



<div class="viewcode-block" id="VarCachingFunctionalAnnotator">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VarCachingFunctionalAnnotator">[docs]</a>
<span class="k">class</span> <span class="nc">VarCachingFunctionalAnnotator</span><span class="p">(</span><span class="n">FunctionalAnnotator</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A class that retrieves a Variant object if it exists or will run the fallback Fuctional Annotator if it does not exist.</span>

<span class="sd">    Methods:</span>
<span class="sd">        annotate(variant_coordinates:VariantCoordinates): Gets data and returns a Variant object for given variant coordinates</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cache</span><span class="p">:</span> <span class="n">VariantAnnotationCache</span><span class="p">,</span> <span class="n">fallback</span><span class="p">:</span> <span class="n">FunctionalAnnotator</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Constructs all necessary attributes for a VarCachingFunctionalAnnotator object</span>

<span class="sd">        Args:</span>
<span class="sd">            cache (VariantAnnotationCache): A VariantAnnotationCache object</span>
<span class="sd">            fallback (FunctionalAnnotator): A FunctionalAnnotator object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_cache</span> <span class="o">=</span> <span class="n">hpotk</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">validate_instance</span><span class="p">(</span><span class="n">cache</span><span class="p">,</span> <span class="n">VariantAnnotationCache</span><span class="p">,</span> <span class="s1">&#39;cache&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fallback</span> <span class="o">=</span> <span class="n">hpotk</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">validate_instance</span><span class="p">(</span><span class="n">fallback</span><span class="p">,</span> <span class="n">FunctionalAnnotator</span><span class="p">,</span> <span class="s1">&#39;fallback&#39;</span><span class="p">)</span>

<div class="viewcode-block" id="VarCachingFunctionalAnnotator.annotate">
<a class="viewcode-back" href="../../../apidocs/genophenocorr.preprocessing.html#genophenocorr.preprocessing.VarCachingFunctionalAnnotator.annotate">[docs]</a>
    <span class="k">def</span> <span class="nf">annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">variant_coordinates</span><span class="p">:</span> <span class="n">VariantCoordinates</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Variant</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Gets Variant for given variant coordinates</span>

<span class="sd">        Args:</span>
<span class="sd">            variant_coordinates (VariantCoordinates): A VariantCoordinates object</span>
<span class="sd">        Returns:</span>
<span class="sd">            Variant: A Variant object</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">hpotk</span><span class="o">.</span><span class="n">util</span><span class="o">.</span><span class="n">validate_instance</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">,</span> <span class="n">VariantCoordinates</span><span class="p">,</span> <span class="s1">&#39;variant_coordinates&#39;</span><span class="p">)</span>
        <span class="n">annotations</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cache</span><span class="o">.</span><span class="n">get_annotations</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">annotations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="c1"># we had cached annotations</span>
            <span class="k">return</span> <span class="n">annotations</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">ann</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fallback</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_cache</span><span class="o">.</span><span class="n">store_annotations</span><span class="p">(</span><span class="n">variant_coordinates</span><span class="p">,</span> <span class="n">ann</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">ann</span></div>
</div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>