<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>genophenocorr.preprocessing package &mdash; genophenocorr 0.1.1dev documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/genophenocorr.css?v=a9fa5eb9" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=8427294d"></script>
        <script src="../_static/doctools.js?v=9a2dae69"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../_static/copybutton.js?v=6dbb43f8"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="genophenocorr.view package" href="genophenocorr.view.html" />
    <link rel="prev" title="genophenocorr.model package" href="genophenocorr.model.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            genophenocorr
          </a>
              <div class="version">
                0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Installation &amp; Tutorial</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../user-guide/index.html">User guide</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="modules.html">API reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="genophenocorr.html">genophenocorr package</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="genophenocorr.html#subpackages">Subpackages</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="genophenocorr.analysis.html">genophenocorr.analysis package</a></li>
<li class="toctree-l4"><a class="reference internal" href="genophenocorr.constants.html">genophenocorr.constants package</a></li>
<li class="toctree-l4"><a class="reference internal" href="genophenocorr.data.html">genophenocorr.data package</a></li>
<li class="toctree-l4"><a class="reference internal" href="genophenocorr.model.html">genophenocorr.model package</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">genophenocorr.preprocessing package</a></li>
<li class="toctree-l4"><a class="reference internal" href="genophenocorr.view.html">genophenocorr.view package</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">genophenocorr</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="modules.html">API reference</a></li>
          <li class="breadcrumb-item"><a href="genophenocorr.html">genophenocorr package</a></li>
      <li class="breadcrumb-item active">genophenocorr.preprocessing package</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/apidocs/genophenocorr.preprocessing.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-genophenocorr.preprocessing">
<span id="genophenocorr-preprocessing-package"></span><h1>genophenocorr.preprocessing package<a class="headerlink" href="#module-genophenocorr.preprocessing" title="Link to this heading"></a></h1>
<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.configure_caching_patient_creator">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">configure_caching_patient_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hpo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.ontology.html#hpotk.ontology.MinimalOntology" title="(in hpo-toolkit)"><span class="pre">MinimalOntology</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'GRCh38.p13'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation_runner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.validate.html#hpotk.validate.ValidationRunner" title="(in hpo-toolkit)"><span class="pre">ValidationRunner</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'VEP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing._phenopacket.PhenopacketPatientCreator"><span class="pre">PhenopacketPatientCreator</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#configure_caching_patient_creator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.configure_caching_patient_creator" title="Link to this definition"></a></dt>
<dd><p>A convenience function for configuring a caching <a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing.PhenopacketPatientCreator"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.preprocessing.PhenopacketPatientCreator</span></code></a>.</p>
<p>To create the patient creator, we need hpo-toolkit’s representation of HPO. Other options are optional.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hpo</strong> – a HPO instance.</p></li>
<li><p><strong>genome_build</strong> – name of the genome build to use, choose from <cite>{‘GRCh37.p13’, ‘GRCh38.p13’}</cite>.</p></li>
<li><p><strong>validation_runner</strong> – an instance of the validation runner.</p></li>
<li><p><strong>cache_dir</strong> – path to the folder where we will cache the results fetched from the remote APIs or <cite>None</cite>
if the data should be cached in <cite>.cache</cite> folder in the current working directory.
In any case, the directory will be created if it does not exist (including non-existing parents).</p></li>
<li><p><strong>variant_fallback</strong> – the fallback variant annotator to use if we cannot find the annotation locally.
Choose from <code class="docutils literal notranslate"><span class="pre">{'VEP'}</span></code> (just one fallback implementation is available at the moment).</p></li>
<li><p><strong>timeout</strong> – timeout in seconds for the VEP API</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.configure_patient_creator">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">configure_patient_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hpo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.ontology.html#hpotk.ontology.MinimalOntology" title="(in hpo-toolkit)"><span class="pre">MinimalOntology</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'GRCh38.p13'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation_runner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.validate.html#hpotk.validate.ValidationRunner" title="(in hpo-toolkit)"><span class="pre">ValidationRunner</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'VEP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'lenient'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing._phenopacket.PhenopacketPatientCreator"><span class="pre">PhenopacketPatientCreator</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#configure_patient_creator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.configure_patient_creator" title="Link to this definition"></a></dt>
<dd><blockquote>
<div><p>^^^ none, lenient, strict -
none = run unless unrunnable
lenient = fix what we can, abort unfixable
strict = abort at any issue</p>
</div></blockquote>
<p>A convenience function for configuring a non-caching <a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing.PhenopacketPatientCreator"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.preprocessing.PhenopacketPatientCreator</span></code></a>.</p>
<p>To create the patient creator, we need hpo-toolkit’s representation of HPO. Other options are optional</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hpo</strong> – a HPO instance.</p></li>
<li><p><strong>genome_build</strong> – name of the genome build to use, choose from <cite>{‘GRCh37.p13’, ‘GRCh38.p13’}</cite>.</p></li>
<li><p><strong>validation_runner</strong> – an instance of the validation runner.
if the data should be cached in <cite>.cache</cite> folder in the current working directory.
In any case, the directory will be created if it does not exist (including non-existing parents).</p></li>
<li><p><strong>variant_fallback</strong> – the fallback variant annotator to use if we cannot find the annotation locally.
Choose from <code class="docutils literal notranslate"><span class="pre">{'VEP'}</span></code> (just one fallback implementation is available at the moment).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.configure_caching_cohort_creator">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">configure_caching_cohort_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hpo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.ontology.html#hpotk.ontology.MinimalOntology" title="(in hpo-toolkit)"><span class="pre">MinimalOntology</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'GRCh38.p13'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation_runner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.validate.html#hpotk.validate.ValidationRunner" title="(in hpo-toolkit)"><span class="pre">ValidationRunner</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cache_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'VEP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genophenocorr.preprocessing.CohortCreator" title="genophenocorr.preprocessing._patient.CohortCreator"><span class="pre">CohortCreator</span></a><span class="p"><span class="pre">[</span></span><span class="pre">Phenopacket</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#configure_caching_cohort_creator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.configure_caching_cohort_creator" title="Link to this definition"></a></dt>
<dd><p>A convenience function for configuring a caching <a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing.PhenopacketPatientCreator"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.preprocessing.PhenopacketPatientCreator</span></code></a>.</p>
<p>To create the patient creator, we need hpo-toolkit’s representation of HPO. Other options are optional.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hpo</strong> – a HPO instance.</p></li>
<li><p><strong>genome_build</strong> – name of the genome build to use, choose from <cite>{‘GRCh37.p13’, ‘GRCh38.p13’}</cite>.</p></li>
<li><p><strong>validation_runner</strong> – an instance of the validation runner.</p></li>
<li><p><strong>cache_dir</strong> – path to the folder where we will cache the results fetched from the remote APIs or <cite>None</cite>
if the data should be cached in <cite>.cache</cite> folder in the current working directory.
In any case, the directory will be created if it does not exist (including non-existing parents).</p></li>
<li><p><strong>variant_fallback</strong> – the fallback variant annotator to use if we cannot find the annotation locally.
Choose from <code class="docutils literal notranslate"><span class="pre">{'VEP'}</span></code> (just one fallback implementation is available at the moment).</p></li>
<li><p><strong>timeout</strong> – timeout in seconds for the VEP API</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.configure_cohort_creator">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">configure_cohort_creator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hpo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.ontology.html#hpotk.ontology.MinimalOntology" title="(in hpo-toolkit)"><span class="pre">MinimalOntology</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'GRCh38.p13'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation_runner</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.validate.html#hpotk.validate.ValidationRunner" title="(in hpo-toolkit)"><span class="pre">ValidationRunner</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variant_fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'VEP'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genophenocorr.preprocessing.CohortCreator" title="genophenocorr.preprocessing._patient.CohortCreator"><span class="pre">CohortCreator</span></a><span class="p"><span class="pre">[</span></span><span class="pre">Phenopacket</span><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#configure_cohort_creator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.configure_cohort_creator" title="Link to this definition"></a></dt>
<dd><p>A convenience function for configuring a non-caching <a class="reference internal" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="genophenocorr.preprocessing.PhenopacketPatientCreator"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.preprocessing.PhenopacketPatientCreator</span></code></a>.</p>
<p>To create the patient creator, we need hpo-toolkit’s representation of HPO. Other options are optional</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hpo</strong> – a HPO instance.</p></li>
<li><p><strong>genome_build</strong> – name of the genome build to use, choose from <cite>{‘GRCh37.p13’, ‘GRCh38.p13’}</cite>.</p></li>
<li><p><strong>validation_runner</strong> – an instance of the validation runner.
if the data should be cached in <cite>.cache</cite> folder in the current working directory.
In any case, the directory will be created if it does not exist (including non-existing parents).</p></li>
<li><p><strong>variant_fallback</strong> – the fallback variant annotator to use if we cannot find the annotation locally.
Choose from <code class="docutils literal notranslate"><span class="pre">{'VEP'}</span></code> (just one fallback implementation is available at the moment).</p></li>
<li><p><strong>timeout</strong> – timeout in seconds for the VEP API</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VariantCoordinateFinder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VariantCoordinateFinder</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#VariantCoordinateFinder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VariantCoordinateFinder" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Generic" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Generic</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code>]</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VariantCoordinateFinder.find_coordinates">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">find_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">T</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.12)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Genotype" title="genophenocorr.model._gt.Genotype"><span class="pre">Genotype</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#VariantCoordinateFinder.find_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VariantCoordinateFinder.find_coordinates" title="Link to this definition"></a></dt>
<dd><p>Determine <code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinates</span></code> and <code class="xref py py-class docutils literal notranslate"><span class="pre">Genotype</span></code> from an <cite>item</cite> of some sort.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – if there is an error of any kind.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.FunctionalAnnotator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">FunctionalAnnotator</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#FunctionalAnnotator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.FunctionalAnnotator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.FunctionalAnnotator.annotate">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#FunctionalAnnotator.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.FunctionalAnnotator.annotate" title="Link to this definition"></a></dt>
<dd><p>Compute functional annotations for the variant coordinates. The annotations can be empty.</p>
<p>Returns: a sequence of transcript annotations
:raises ValueError if the annotation cannot proceed due to the remote resource being offline, etc.:</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProteinMetadataService">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">ProteinMetadataService</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#ProteinMetadataService"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProteinMetadataService" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A service for obtaining annotations for a given protein accession ID.</p>
<p>The annotations include elements of the <code class="xref py py-class docutils literal notranslate"><span class="pre">ProteinMetadata</span></code> class.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProteinMetadataService.annotate">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_api.html#ProteinMetadataService.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProteinMetadataService.annotate" title="Link to this definition"></a></dt>
<dd><p>Prepare <cite>ProteinMetadata</cite> for a protein with given <cite>protein_id</cite> accession ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>string</em>) – A accession ID <cite>str</cite> (e.g. <cite>NP_001027558.1</cite>)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a <cite>ProteinMetadata</cite> container with the protein metadata</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model.ProteinMetadata">ProteinMetadata</a></p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – </p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PatientCreator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">PatientCreator</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_patient.html#PatientCreator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PatientCreator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Generic" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Generic</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code>], <a class="reference internal" href="#genophenocorr.preprocessing.Auditor" title="genophenocorr.preprocessing._audit.Auditor"><code class="xref py py-class docutils literal notranslate"><span class="pre">Auditor</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code>, <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Patient" title="genophenocorr.model._cohort.Patient"><code class="xref py py-class docutils literal notranslate"><span class="pre">Patient</span></code></a>]</p>
<p><cite>PatientCreator</cite> can create a <cite>Patient</cite> from some input <cite>T</cite>.</p>
<p><cite>PatientCreator</cite> is an <cite>Auditor</cite>, hence the input is sanitized and any errors are reported to the caller.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.CohortCreator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">CohortCreator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">patient_creator</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.PatientCreator" title="genophenocorr.preprocessing._patient.PatientCreator"><span class="pre">PatientCreator</span></a><span class="p"><span class="pre">[</span></span><span class="pre">T</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_patient.html#CohortCreator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.CohortCreator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Generic" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Generic</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code>], <a class="reference internal" href="#genophenocorr.preprocessing.Auditor" title="genophenocorr.preprocessing._audit.Auditor"><code class="xref py py-class docutils literal notranslate"><span class="pre">Auditor</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Iterable</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">T</span></code>], <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Cohort" title="genophenocorr.model._cohort.Cohort"><code class="xref py py-class docutils literal notranslate"><span class="pre">Cohort</span></code></a>]</p>
<p><cite>CohortCreator</cite> creates a cohort from an iterable of some <cite>T</cite> where <cite>T</cite> represents a cohort member.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.CohortCreator.process">
<span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inputs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.12)"><span class="pre">Iterable</span></a><span class="p"><span class="pre">[</span></span><span class="pre">T</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">notepad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing._audit.Notepad"><span class="pre">Notepad</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Cohort" title="genophenocorr.model._cohort.Cohort"><span class="pre">Cohort</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_patient.html#CohortCreator.process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.CohortCreator.process" title="Link to this definition"></a></dt>
<dd><p>Audit and sanitize the <cite>data</cite>, record the issues to the <cite>notepad</cite> and return the sanitized data.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">PhenopacketVariantCoordinateFinder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">GenomeBuild</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hgvs_coordinate_finder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.VariantCoordinateFinder" title="genophenocorr.preprocessing._api.VariantCoordinateFinder"><span class="pre">VariantCoordinateFinder</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenopacket.html#PhenopacketVariantCoordinateFinder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.VariantCoordinateFinder" title="genophenocorr.preprocessing._api.VariantCoordinateFinder"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinateFinder</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicInterpretation</span></code>]</p>
<p><cite>PhenopacketVariantCoordinateFinder</cite> figures out <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.model.VariantCoordinates</span></code></a>
and <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Genotype" title="genophenocorr.model.Genotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.model.Genotype</span></code></a> from <cite>GenomicInterpretation</cite> element of Phenopacket Schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>build</strong> – genome build to use in <a href="#id1"><span class="problematic" id="id2">`</span></a>VariantCoordinates</p></li>
<li><p><strong>hgvs_coordinate_finder</strong> – the coordinate finder to use for parsing HGVS expressions</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder.find_coordinates">
<span class="sig-name descname"><span class="pre">find_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">GenomicInterpretation</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.12)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Genotype" title="genophenocorr.model._gt.Genotype"><span class="pre">Genotype</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenopacket.html#PhenopacketVariantCoordinateFinder.find_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenopacketVariantCoordinateFinder.find_coordinates" title="Link to this definition"></a></dt>
<dd><p>Creates a VariantCoordinates object from the data in a given Phenopacket</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>item</strong> (<em>GenomicInterpretation</em>) – a genomic interpretation element from Phenopacket Schema</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple of <code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinates</span></code> and <code class="xref py py-class docutils literal notranslate"><span class="pre">Genotype</span></code></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#tuple" title="(in Python v3.12)">tuple</a>[VariantCall, <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Genotype" title="genophenocorr.model.Genotype">Genotype</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenopacketPatientCreator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">PhenopacketPatientCreator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">GenomeBuild</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">phenotype_creator</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.PhenotypeCreator" title="genophenocorr.preprocessing._phenotype.PhenotypeCreator"><span class="pre">PhenotypeCreator</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">var_func_ann</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.FunctionalAnnotator" title="genophenocorr.preprocessing._api.FunctionalAnnotator"><span class="pre">FunctionalAnnotator</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">hgvs_coordinate_finder</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.VariantCoordinateFinder" title="genophenocorr.preprocessing._api.VariantCoordinateFinder"><span class="pre">VariantCoordinateFinder</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenopacket.html#PhenopacketPatientCreator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenopacketPatientCreator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.PatientCreator" title="genophenocorr.preprocessing._patient.PatientCreator"><code class="xref py py-class docutils literal notranslate"><span class="pre">PatientCreator</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Phenopacket</span></code>]</p>
<p><cite>PhenopacketPatientCreator</cite> transforms <cite>Phenopacket</cite> into <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Patient" title="genophenocorr.model.Patient"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.model.Patient</span></code></a>.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenopacketPatientCreator.process">
<span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inputs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Phenopacket</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">notepad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing._audit.Notepad"><span class="pre">Notepad</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Patient" title="genophenocorr.model._cohort.Patient"><span class="pre">Patient</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenopacket.html#PhenopacketPatientCreator.process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenopacketPatientCreator.process" title="Link to this definition"></a></dt>
<dd><p>Creates a Patient from the data in a given Phenopacket</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>inputs</strong> (<em>Phenopacket</em>) – A Phenopacket object</p></li>
<li><p><strong>notepad</strong> (<a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing.Notepad"><em>Notepad</em></a>) – notepad to write down the issues</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Patient object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Patient" title="genophenocorr.model.Patient">Patient</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.load_phenopacket_folder">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">load_phenopacket_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pp_directory</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">cohort_creator</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.CohortCreator" title="genophenocorr.preprocessing._patient.CohortCreator"><span class="pre">CohortCreator</span></a><span class="p"><span class="pre">[</span></span><span class="pre">Phenopacket</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">validation_policy</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Literal" title="(in Python v3.12)"><span class="pre">Literal</span></a><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'none'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'lenient'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'strict'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'none'</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Cohort" title="genophenocorr.model._cohort.Cohort"><span class="pre">Cohort</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#load_phenopacket_folder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.load_phenopacket_folder" title="Link to this definition"></a></dt>
<dd><p>Creates a Patient object for each phenopacket formatted JSON file in the given directory <cite>pp_directory</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pp_directory</strong> – path to a folder with phenopacket JSON files. An error is raised if the path does not point to
a directory with at least one phenopacket.</p></li>
<li><p><strong>cohort_creator</strong> – cohort creator for turning a sequence of phenopacket into a <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Cohort" title="genophenocorr.model.Cohort"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.model.Cohort</span></code></a>.</p></li>
<li><p><strong>validation_policy</strong> – a <cite>str</cite> with the validation policy. The value must be one of <cite>{‘none’, ‘lenient’, ‘strict’}</cite></p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a cohort made of the phenopackets</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.load_phenopacket">
<span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">load_phenopacket</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">phenopacket_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Phenopacket</span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_config.html#load_phenopacket"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.load_phenopacket" title="Link to this definition"></a></dt>
<dd><p>Load phenopacket JSON file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>phenopacket_path</strong> – a <cite>str</cite> pointing to phenopacket JSON file.</p>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenotypeCreator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">PhenotypeCreator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hpo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.ontology.html#hpotk.ontology.MinimalOntology" title="(in hpo-toolkit)"><span class="pre">MinimalOntology</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">validator</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://ielis.github.io/hpo-toolkit/latest/apidocs/hpotk.validate.html#hpotk.validate.ValidationRunner" title="(in hpo-toolkit)"><span class="pre">ValidationRunner</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenotype.html#PhenotypeCreator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenotypeCreator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.Auditor" title="genophenocorr.preprocessing._audit.Auditor"><code class="xref py py-class docutils literal notranslate"><span class="pre">Auditor</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Iterable</span></code></a>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Tuple</span></code>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>, <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code></a>]], <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></a>[<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Phenotype" title="genophenocorr.model._phenotype.Phenotype"><code class="xref py py-class docutils literal notranslate"><span class="pre">Phenotype</span></code></a>]]</p>
<p><cite>PhenotypeCreator</cite> validates the input phenotype features and prepares them for the downstream analysis.</p>
<p>The creator expects an iterable with tuples that contain a CURIE and status. The CURIE must correspond
to a HPO term identifier and status must be a <cite>bool</cite>.</p>
<p>The creator prunes CURIES with simple errors such as malformed CURIE or non-HPO terms and validates the rest
with HPO toolkit’s validator.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.PhenotypeCreator.process">
<span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inputs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterable" title="(in Python v3.12)"><span class="pre">Iterable</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Tuple" title="(in Python v3.12)"><span class="pre">Tuple</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">notepad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing._audit.Notepad"><span class="pre">Notepad</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Phenotype" title="genophenocorr.model._phenotype.Phenotype"><span class="pre">Phenotype</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_phenotype.html#PhenotypeCreator.process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.PhenotypeCreator.process" title="Link to this definition"></a></dt>
<dd><p>Creates a list of Phenotype objects from term IDs and checks if the term IDs satisfy the validation requirements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>inputs</strong> (<em>Iterable</em><em>[</em><em>Tuple</em><em>[</em><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><em>str</em></a><em>, </em><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a><em>]</em><em>]</em>) – A list of Tuples, structured (HPO IDs, boolean- True if observed)</p></li>
<li><p><strong>notepad</strong> – Node</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A sequence of Phenotype objects</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProteinAnnotationCache">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">ProteinAnnotationCache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datadir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_protein.html#ProteinAnnotationCache"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProteinAnnotationCache" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A class that stores or retrieves ProteinMetadata objects using pickle format</p>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_annotations(protein_id</span></span></dt>
<dd><p>str): Searches a given data directory for a pickle file with given ID and returns ProteinMetadata</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">store_annotations(protein_id</span></span></dt>
<dd><p>str, annotation:Sequence[ProteinMetadata]): Creates a pickle file with given ID and stores the given ProteinMetadata into that file</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProteinAnnotationCache.get_annotations">
<span class="sig-name descname"><span class="pre">get_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_protein.html#ProteinAnnotationCache.get_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProteinAnnotationCache.get_annotations" title="Link to this definition"></a></dt>
<dd><p>Searches a given data directory for a pickle file with given ID and returns ProteinMetadata from file. Returns None if no file is found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>string</em>) – The protein_id associated with the desired ProteinMetadata</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProteinAnnotationCache.store_annotations">
<span class="sig-name descname"><span class="pre">store_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_protein.html#ProteinAnnotationCache.store_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProteinAnnotationCache.store_annotations" title="Link to this definition"></a></dt>
<dd><p>Creates a pickle file with the given protein id in the file name. Loads the ProteinMetadata given into the file for storage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>protein_id</strong> (<em>string</em>) – The protein_id associated with the ProteinMetadata</p></li>
<li><p><strong>annotation</strong> (<em>Sequence</em><em>[</em><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model.ProteinMetadata"><em>ProteinMetadata</em></a><em>]</em>) – A sequence of ProteinMetadata objects that will be stored under the given protein id</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProtCachingMetadataService">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">ProtCachingMetadataService</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.ProteinAnnotationCache" title="genophenocorr.preprocessing._protein.ProteinAnnotationCache"><span class="pre">ProteinAnnotationCache</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.ProteinMetadataService" title="genophenocorr.preprocessing._api.ProteinMetadataService"><span class="pre">ProteinMetadataService</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_protein.html#ProtCachingMetadataService"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProtCachingMetadataService" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.ProteinMetadataService" title="genophenocorr.preprocessing._api.ProteinMetadataService"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProteinMetadataService</span></code></a></p>
<p>A class that retrieves ProteinMetadata if it exists or will run the fallback Fuctional Annotator if it does not exist.</p>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">annotate(protein_id</span></span></dt>
<dd><p>str): Gets metadata and returns ProteinMetadata for given protein ID</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.ProtCachingMetadataService.annotate">
<span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_protein.html#ProtCachingMetadataService.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.ProtCachingMetadataService.annotate" title="Link to this definition"></a></dt>
<dd><p>Gets metadata for given protein ID</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>string</em>) – A protein ID</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A ProteinMetadata object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model.ProteinMetadata">ProteinMetadata</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.UniprotProteinMetadataService">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">UniprotProteinMetadataService</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_uniprot.html#UniprotProteinMetadataService"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.UniprotProteinMetadataService" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.ProteinMetadataService" title="genophenocorr.preprocessing._api.ProteinMetadataService"><code class="xref py py-class docutils literal notranslate"><span class="pre">ProteinMetadataService</span></code></a></p>
<p>A class that creates ProteinMetadata objects from data found with the Uniprot REST API.
More info on the Uniprot REST API here - <a class="reference external" href="https://www.uniprot.org/help/programmatic_access">https://www.uniprot.org/help/programmatic_access</a></p>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">annotate(protein_id</span></span></dt>
<dd><p>str): Gets metadata and creates ProteinMetadata for given protein ID</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.UniprotProteinMetadataService.parse_uniprot_json">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">parse_uniprot_json</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">payload</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_uniprot.html#UniprotProteinMetadataService.parse_uniprot_json"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.UniprotProteinMetadataService.parse_uniprot_json" title="Link to this definition"></a></dt>
<dd><p>Try to extract <cite>ProteinMetadata</cite> corresponding to <cite>protein_id</cite> from the Uniprot JSON <cite>payload</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>payload</strong> – a JSON object corresponding to Uniprot response</p></li>
<li><p><strong>protein_id</strong> – a <cite>str</cite> with the accession the protein of interest</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>a complete instance of <cite>ProteinMetadata</cite></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/exceptions.html#ValueError" title="(in Python v3.12)"><strong>ValueError</strong></a> – </p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.UniprotProteinMetadataService.annotate">
<span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">protein_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model._protein.ProteinMetadata"><span class="pre">ProteinMetadata</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_uniprot.html#UniprotProteinMetadataService.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.UniprotProteinMetadataService.annotate" title="Link to this definition"></a></dt>
<dd><p>Get metadata for given protein ID. This class specifically only works with a RefSeq database ID (e.g. NP_037407.4)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>protein_id</strong> (<em>string</em>) – A protein ID</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A sequence of ProteinMetadata objects, or an empty sequence if no data was found.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sequence[<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.ProteinMetadata" title="genophenocorr.model.ProteinMetadata">ProteinMetadata</a>]</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VepFunctionalAnnotator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_computational_txs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vep.html#VepFunctionalAnnotator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.FunctionalAnnotator" title="genophenocorr.preprocessing._api.FunctionalAnnotator"><code class="xref py py-class docutils literal notranslate"><span class="pre">FunctionalAnnotator</span></code></a></p>
<p>A <cite>FunctionalAnnotator</cite> that uses Variant Effect Predictor (VEP) REST API to
do functional variant annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>include_computational_txs</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><em>bool</em></a>) – Include computational transcripts, such as</p></li>
<li><p><strong>XM_.</strong> (<em>RefSeq</em>)</p></li>
<li><p><strong>timeout</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><em>int</em></a>) – Timeout in seconds</p></li>
</ul>
</dd>
</dl>
<dl class="py attribute">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator.NONCODING_EFFECTS">
<span class="sig-name descname"><span class="pre">NONCODING_EFFECTS</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{VariantEffect.DOWNSTREAM_GENE_VARIANT,</span> <span class="pre">VariantEffect.FIVE_PRIME_UTR_VARIANT,</span> <span class="pre">VariantEffect.INTERGENIC_VARIANT,</span> <span class="pre">VariantEffect.INTRON_VARIANT,</span> <span class="pre">VariantEffect.NON_CODING_TRANSCRIPT_EXON_VARIANT,</span> <span class="pre">VariantEffect.NON_CODING_TRANSCRIPT_VARIANT,</span> <span class="pre">VariantEffect.SPLICE_ACCEPTOR_VARIANT,</span> <span class="pre">VariantEffect.SPLICE_DONOR_5TH_BASE_VARIANT,</span> <span class="pre">VariantEffect.SPLICE_DONOR_VARIANT,</span> <span class="pre">VariantEffect.SPLICE_POLYPYRIMIDINE_TRACT_VARIANT,</span> <span class="pre">VariantEffect.THREE_PRIME_UTR_VARIANT,</span> <span class="pre">VariantEffect.UPSTREAM_GENE_VARIANT}</span></em><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator.NONCODING_EFFECTS" title="Link to this definition"></a></dt>
<dd><p>Non-coding variant effects where we do not complain if the functional annotation lacks the protein effects.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator.annotate">
<span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vep.html#VepFunctionalAnnotator.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator.annotate" title="Link to this definition"></a></dt>
<dd><p>Perform functional annotation using Variant Effect Predictor (VEP) REST API.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>variant_coordinates</strong> (<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><em>VariantCoordinates</em></a>) – A VariantCoordinates object</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A sequence of transcript
annotations for the variant coordinates</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><em>Sequence</em></a>[<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model.TranscriptAnnotation">TranscriptAnnotation</a>]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError if VEP times out</strong><strong> or </strong><strong>does not return a response</strong><strong> or </strong><strong>if the response is not formatted as we expect.</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator.process_response">
<span class="sig-name descname"><span class="pre">process_response</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_key</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">response</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vep.html#VepFunctionalAnnotator.process_response"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator.process_response" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator.fetch_response">
<span class="sig-name descname"><span class="pre">fetch_response</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Mapping" title="(in Python v3.12)"><span class="pre">Mapping</span></a><span class="p"><span class="pre">[</span></span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Any" title="(in Python v3.12)"><span class="pre">Any</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vep.html#VepFunctionalAnnotator.fetch_response"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator.fetch_response" title="Link to this definition"></a></dt>
<dd><p>Get a <cite>dict</cite> with the response from the VEP REST API.
:param variant_coordinates: a query <code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinates</span></code>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VepFunctionalAnnotator.format_coordinates_for_vep_query">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">format_coordinates_for_vep_query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">vc</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vep.html#VepFunctionalAnnotator.format_coordinates_for_vep_query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VepFunctionalAnnotator.format_coordinates_for_vep_query" title="Link to this definition"></a></dt>
<dd><p>Converts the 0-based VariantCoordinates to ones that will be interpreted
correctly by VEP</p>
<p>Example - an insertion/duplication of G after the given G at coordinate 3:
1 2 3 4 5
A C G T A</p>
<p>0-based: 2 3 G GG       1-based: 3 G GG         VEP: 4 3 - G</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>vc</strong> (<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><em>VariantCoordinates</em></a>) – A VariantCoordinates object</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The variant coordinates formatted to work with VEP</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>string</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VariantAnnotationCache">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VariantAnnotationCache</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">datadir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_variant.html#VariantAnnotationCache"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VariantAnnotationCache" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>A class that stores or retrieves Variant objects using pickle format</p>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_annotations(variant_coordinates</span></span></dt>
<dd><p>VariantCoordinates): Searches a given data directory for a pickle file with variant coordinates and returns a Variant object</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">store_annotations(variant_coordinates</span></span></dt>
<dd><p>VariantCoordinates, annotation:Variant): Creates a pickle file with variant coordinates and stores the given Variant object into that file</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VariantAnnotationCache.get_annotations">
<span class="sig-name descname"><span class="pre">get_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_variant.html#VariantAnnotationCache.get_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VariantAnnotationCache.get_annotations" title="Link to this definition"></a></dt>
<dd><p>Searches a given data directory for a pickle file with given variant coordinates and returns Variant from file. Returns None if no file is found.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>variant_coordinates</strong> (<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><em>VariantCoordinates</em></a>) – The variant_coordinates associated with the desired Variant</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VariantAnnotationCache.store_annotations">
<span class="sig-name descname"><span class="pre">store_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_variant.html#VariantAnnotationCache.store_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VariantAnnotationCache.store_annotations" title="Link to this definition"></a></dt>
<dd><p>Creates a pickle file with the given variant coordinates in the file name. Loads the Variant object given into the file for storage.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_coordinates</strong> (<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><em>VariantCoordinates</em></a>) – The variant_coordinates associated with the desired Variant</p></li>
<li><p><strong>annotations</strong> (<a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><em>Sequence</em></a><em>[</em><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model.TranscriptAnnotation"><em>TranscriptAnnotation</em></a><em>]</em>) – Annotations that will be stored under the given variant coordinates</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VarCachingFunctionalAnnotator">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VarCachingFunctionalAnnotator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cache</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.VariantAnnotationCache" title="genophenocorr.preprocessing._variant.VariantAnnotationCache"><span class="pre">VariantAnnotationCache</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">fallback</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.FunctionalAnnotator" title="genophenocorr.preprocessing._api.FunctionalAnnotator"><span class="pre">FunctionalAnnotator</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_variant.html#VarCachingFunctionalAnnotator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VarCachingFunctionalAnnotator" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.FunctionalAnnotator" title="genophenocorr.preprocessing._api.FunctionalAnnotator"><code class="xref py py-class docutils literal notranslate"><span class="pre">FunctionalAnnotator</span></code></a></p>
<p>A class that retrieves a Variant object if it exists or will run the fallback Fuctional Annotator if it does not exist.</p>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">annotate(variant_coordinates</span></span></dt>
<dd><p>VariantCoordinates): Gets data and returns a Variant object for given variant coordinates</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VarCachingFunctionalAnnotator.annotate">
<span class="sig-name descname"><span class="pre">annotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">variant_coordinates</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model._variant.TranscriptAnnotation"><span class="pre">TranscriptAnnotation</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_variant.html#VarCachingFunctionalAnnotator.annotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VarCachingFunctionalAnnotator.annotate" title="Link to this definition"></a></dt>
<dd><p>Gets Variant for given variant coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>variant_coordinates</strong> (<a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model.VariantCoordinates"><em>VariantCoordinates</em></a>) – A VariantCoordinates object</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A Variant object</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.Variant" title="genophenocorr.model.Variant">Variant</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVHgvsVariantCoordinateFinder">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VVHgvsVariantCoordinateFinder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">GenomeBuild</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVHgvsVariantCoordinateFinder"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVHgvsVariantCoordinateFinder" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.VariantCoordinateFinder" title="genophenocorr.preprocessing._api.VariantCoordinateFinder"><code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinateFinder</span></code></a>[<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></a>]</p>
<p><cite>VVHgvsVariantCoordinateFinder</cite> uses Variant Validator’s REST API to build <code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinates</span></code>
from an HGVS string.</p>
<p>The finder takes an HGVS <cite>str</cite> (e.g. <cite>NM_005912.3:c.253A&gt;G</cite>) and extracts the variant coordinates from the response.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_build</strong> – the genome build to use to construct <code class="xref py py-class docutils literal notranslate"><span class="pre">VariantCoordinates</span></code></p></li>
<li><p><strong>timeout</strong> – the REST API request timeout</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVHgvsVariantCoordinateFinder.find_coordinates">
<span class="sig-name descname"><span class="pre">find_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.VariantCoordinates" title="genophenocorr.model._variant.VariantCoordinates"><span class="pre">VariantCoordinates</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVHgvsVariantCoordinateFinder.find_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVHgvsVariantCoordinateFinder.find_coordinates" title="Link to this definition"></a></dt>
<dd><p>Extracts variant coordinates from an HGVS string using Variant Validator’s REST API.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>item</strong> – a hgvs string</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>variant coordinates</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVTranscriptCoordinateService">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">VVTranscriptCoordinateService</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_build</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">GenomeBuild</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.12)"><span class="pre">float</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">30.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVTranscriptCoordinateService"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVTranscriptCoordinateService" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">TranscriptCoordinateService</span></code></p>
<p><cite>VVTranscriptCoordinateService</cite> fetches the transcript coordinates from the Variant Validator REST API.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_build</strong> – the genome build for constructing the transcript coordinates.</p></li>
<li><p><strong>timeout</strong> – a positive <cite>float</cite> with the REST API timeout in seconds.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVTranscriptCoordinateService.fetch">
<span class="sig-name descname"><span class="pre">fetch</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptInfoAware" title="genophenocorr.model._variant.TranscriptInfoAware"><span class="pre">TranscriptInfoAware</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptCoordinates" title="genophenocorr.model._tx.TranscriptCoordinates"><span class="pre">TranscriptCoordinates</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVTranscriptCoordinateService.fetch"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVTranscriptCoordinateService.fetch" title="Link to this definition"></a></dt>
<dd><p>Get tx coordinates for a tx ID or an entity that knows about the tx ID.</p>
<p>The method will raise an exception in case of an issue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tx</strong> – a <cite>str</cite> with tx ID (e.g. <cite>NM_002834.5</cite>) or an entity that knows about the transcript ID</p>
</dd>
</dl>
<p>:param (e.g. <a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptAnnotation" title="genophenocorr.model.TranscriptAnnotation"><code class="xref py py-class docutils literal notranslate"><span class="pre">genophenocorr.model.TranscriptAnnotation</span></code></a>).:</p>
<p>Returns: the transcript coordinates.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVTranscriptCoordinateService.get_response">
<span class="sig-name descname"><span class="pre">get_response</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVTranscriptCoordinateService.get_response"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVTranscriptCoordinateService.get_response" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.VVTranscriptCoordinateService.parse_response">
<span class="sig-name descname"><span class="pre">parse_response</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tx_id</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">response</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="genophenocorr.model.html#genophenocorr.model.TranscriptCoordinates" title="genophenocorr.model._tx.TranscriptCoordinates"><span class="pre">TranscriptCoordinates</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_vv.html#VVTranscriptCoordinateService.parse_response"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.VVTranscriptCoordinateService.parse_response" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Auditor">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">Auditor</span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Auditor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Auditor" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Generic" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Generic</span></code></a>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">IN</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">OUT</span></code>]</p>
<p><cite>Auditor</cite> checks the inputs for sanity issues and relates the issues with sanitized inputs
as <code class="xref py py-class docutils literal notranslate"><span class="pre">SanitationResults</span></code>.</p>
<p>The auditor may sanitize the input as a matter of discretion and returns the input as <cite>OUT</cite>.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Auditor.prepare_notepad">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prepare_notepad</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree" title="genophenocorr.preprocessing._audit.NotepadTree"><span class="pre">NotepadTree</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Auditor.prepare_notepad"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Auditor.prepare_notepad" title="Link to this definition"></a></dt>
<dd><p>Prepare a <a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing.Notepad"><code class="xref py py-class docutils literal notranslate"><span class="pre">Notepad</span></code></a> for recording issues and errors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>label</strong> – a <cite>str</cite> with the top-level section label.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>an instance of <a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree" title="genophenocorr.preprocessing.NotepadTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">NotepadTree</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree" title="genophenocorr.preprocessing.NotepadTree">NotepadTree</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Auditor.process">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">process</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">IN</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">notepad</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing._audit.Notepad"><span class="pre">Notepad</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">OUT</span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Auditor.process"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Auditor.process" title="Link to this definition"></a></dt>
<dd><p>Audit and sanitize the <cite>data</cite>, record the issues to the <cite>notepad</cite> and return the sanitized data.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.DataSanityIssue">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">DataSanityIssue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Level" title="genophenocorr.preprocessing._audit.Level"><span class="pre">Level</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">solution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#DataSanityIssue"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.DataSanityIssue" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p><cite>DataSanityIssue</cite> summarizes an issue found in the input data.</p>
<p>The issue has a <cite>level</cite>, a <cite>message</cite> with human-friendly description, and an optional <cite>solution</cite>
for removing the issue.</p>
<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.DataSanityIssue.level">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">level</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#genophenocorr.preprocessing.Level" title="genophenocorr.preprocessing._audit.Level"><span class="pre">Level</span></a></em><a class="headerlink" href="#genophenocorr.preprocessing.DataSanityIssue.level" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.DataSanityIssue.message">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">message</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#genophenocorr.preprocessing.DataSanityIssue.message" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.DataSanityIssue.solution">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">solution</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></em><a class="headerlink" href="#genophenocorr.preprocessing.DataSanityIssue.solution" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Level">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">Level</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">module</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">qualname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">boundary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Level"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Level" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/enum.html#enum.Enum" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Enum</span></code></a></p>
<p>An enum to represent severity of the <a class="reference internal" href="#genophenocorr.preprocessing.DataSanityIssue" title="genophenocorr.preprocessing.DataSanityIssue"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataSanityIssue</span></code></a>.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Level.WARN">
<span class="sig-name descname"><span class="pre">WARN</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1</span></em><a class="headerlink" href="#genophenocorr.preprocessing.Level.WARN" title="Link to this definition"></a></dt>
<dd><p>Warning is an issue when something not entirely right. However, unlike <a class="reference internal" href="#genophenocorr.preprocessing.Level.ERROR" title="genophenocorr.preprocessing.Level.ERROR"><code class="xref py py-class docutils literal notranslate"><span class="pre">Level.ERROR</span></code></a>, the analysis should
complete albeit with sub-optimal results 😧.</p>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Level.ERROR">
<span class="sig-name descname"><span class="pre">ERROR</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">2</span></em><a class="headerlink" href="#genophenocorr.preprocessing.Level.ERROR" title="Link to this definition"></a></dt>
<dd><p>Error is a serious issue in the input data and the downstream analysis may not complete or the analysis results
may be malarkey 😱.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">Notepad</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference external" href="https://docs.python.org/3/library/functions.html#object" title="(in Python v3.12)"><code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></a></p>
<p>Record issues encountered during parsing/validation of a hierarchical data structure.</p>
<p>The issues can be organized in sections. <cite>Notepad</cite> keeps track of issues in one section
and the subsections can be created by calling <a class="reference internal" href="#genophenocorr.preprocessing.Notepad.add_subsection" title="genophenocorr.preprocessing.Notepad.add_subsection"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_subsection()</span></code></a>. The function returns
an instance responsible for issues of a subsection.</p>
<p>A collection of the issues from the current section are available via <a class="reference internal" href="#genophenocorr.preprocessing.Notepad.issues" title="genophenocorr.preprocessing.Notepad.issues"><code class="xref py py-attr docutils literal notranslate"><span class="pre">issues</span></code></a> property
and the convenience functions provide iterators over error and warnings.</p>
<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.add_subsection">
<em class="property"><span class="pre">abstract</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">add_subsection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.add_subsection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.add_subsection" title="Link to this definition"></a></dt>
<dd><p>Add a labeled subsection.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a notepad for recording issues within the subsection.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing.Notepad">Notepad</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.label">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">label</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></em><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.label" title="Link to this definition"></a></dt>
<dd><p>Get a <cite>str</cite> with the section label.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.issues">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">issues</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Sequence" title="(in Python v3.12)"><span class="pre">Sequence</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#genophenocorr.preprocessing.DataSanityIssue" title="genophenocorr.preprocessing._audit.DataSanityIssue"><span class="pre">DataSanityIssue</span></a><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.issues" title="Link to this definition"></a></dt>
<dd><p>Get an iterable with the issues of the current section.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.add_issue">
<span class="sig-name descname"><span class="pre">add_issue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#genophenocorr.preprocessing.Level" title="genophenocorr.preprocessing._audit.Level"><span class="pre">Level</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">solution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.add_issue"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.add_issue" title="Link to this definition"></a></dt>
<dd><p>Add an issue with certain <cite>level</cite>, <cite>message</cite>, and an optional <cite>solution</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.add_error">
<span class="sig-name descname"><span class="pre">add_error</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">solution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.add_error"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.add_error" title="Link to this definition"></a></dt>
<dd><p>A convenience function for adding an <em>error</em> with a <cite>message</cite> and an optional <cite>solution</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.add_warning">
<span class="sig-name descname"><span class="pre">add_warning</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">message</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">solution</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.12)"><span class="pre">None</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.add_warning"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.add_warning" title="Link to this definition"></a></dt>
<dd><p>A convenience function for adding a <em>warning</em> with a <cite>message</cite> and an optional <cite>solution</cite>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.errors">
<span class="sig-name descname"><span class="pre">errors</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterator" title="(in Python v3.12)"><span class="pre">Iterator</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#genophenocorr.preprocessing.DataSanityIssue" title="genophenocorr.preprocessing._audit.DataSanityIssue"><span class="pre">DataSanityIssue</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.errors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.errors" title="Link to this definition"></a></dt>
<dd><p>Iterate over the errors of the current section.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.warnings">
<span class="sig-name descname"><span class="pre">warnings</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/typing.html#typing.Iterator" title="(in Python v3.12)"><span class="pre">Iterator</span></a><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#genophenocorr.preprocessing.DataSanityIssue" title="genophenocorr.preprocessing._audit.DataSanityIssue"><span class="pre">DataSanityIssue</span></a><span class="p"><span class="pre">]</span></span></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.warnings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.warnings" title="Link to this definition"></a></dt>
<dd><p>Iterate over the warnings of the current section.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.error_count">
<span class="sig-name descname"><span class="pre">error_count</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.error_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.error_count" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>count of errors found in this section.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.Notepad.warning_count">
<span class="sig-name descname"><span class="pre">warning_count</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#Notepad.warning_count"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.Notepad.warning_count" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>count of warnings found in this section.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)">int</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">genophenocorr.preprocessing.</span></span><span class="sig-name descname"><span class="pre">NotepadTree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">label</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing._audit.Notepad"><code class="xref py py-class docutils literal notranslate"><span class="pre">Notepad</span></code></a></p>
<p><cite>NotepadTree</cite> implements <a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing.Notepad"><code class="xref py py-class docutils literal notranslate"><span class="pre">Notepad</span></code></a> using a tree where each tree node corresponds to a (sub)section. The node
can have <cite>0..n</cite> children.</p>
<p>Each node has a <code class="xref py py-attr docutils literal notranslate"><span class="pre">label</span></code>, a collection of issues, and children with subsections. For convenience, the node
has <a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree.level" title="genophenocorr.preprocessing.NotepadTree.level"><code class="xref py py-attr docutils literal notranslate"><span class="pre">level</span></code></a> to correspond to the depth of the node within the tree (the level of the root node is <cite>0</cite>).</p>
<p>The nodes can be accessed via <a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree.children" title="genophenocorr.preprocessing.NotepadTree.children"><code class="xref py py-attr docutils literal notranslate"><span class="pre">children</span></code></a> property or through convenience methods for tree traversal, either
using the visitor pattern (<a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree.visit" title="genophenocorr.preprocessing.NotepadTree.visit"><code class="xref py py-func docutils literal notranslate"><span class="pre">visit()</span></code></a>) or by iterating over the nodes via <a class="reference internal" href="#genophenocorr.preprocessing.NotepadTree.iterate_nodes" title="genophenocorr.preprocessing.NotepadTree.iterate_nodes"><code class="xref py py-func docutils literal notranslate"><span class="pre">iterate_nodes()</span></code></a>. In both cases,
the traversal is done in the depth-first fashion.</p>
<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.children">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">children</span></span><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.children" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.level">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">level</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.12)"><span class="pre">int</span></a></em><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.level" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.add_subsection">
<span class="sig-name descname"><span class="pre">add_subsection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.12)"><span class="pre">str</span></a></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.add_subsection"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.add_subsection" title="Link to this definition"></a></dt>
<dd><p>Add a labeled subsection.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>a notepad for recording issues within the subsection.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#genophenocorr.preprocessing.Notepad" title="genophenocorr.preprocessing.Notepad">Notepad</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.visit">
<span class="sig-name descname"><span class="pre">visit</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">visitor</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.visit"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.visit" title="Link to this definition"></a></dt>
<dd><p>Perform a depth-first search on the tree and call <cite>visitor</cite> with all nodes.
:param visitor: a callable that takes the current node as a single argument.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.iterate_nodes">
<span class="sig-name descname"><span class="pre">iterate_nodes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.iterate_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.iterate_nodes" title="Link to this definition"></a></dt>
<dd><p>Iterate over nodes in the depth-first fashion.</p>
<p>Returns: a depth-first node iterator.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.has_warnings">
<span class="sig-name descname"><span class="pre">has_warnings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_subsections</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.has_warnings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.has_warnings" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>True</cite> if one or more warnings were found in the current section or its subsections.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.has_errors">
<span class="sig-name descname"><span class="pre">has_errors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_subsections</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.has_errors"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.has_errors" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>True</cite> if one or more errors were found in the current section or its subsections.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)">bool</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="genophenocorr.preprocessing.NotepadTree.has_errors_or_warnings">
<span class="sig-name descname"><span class="pre">has_errors_or_warnings</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">include_subsections</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)"><span class="pre">bool</span></a></span></span><a class="reference internal" href="../_modules/genophenocorr/preprocessing/_audit.html#NotepadTree.has_errors_or_warnings"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#genophenocorr.preprocessing.NotepadTree.has_errors_or_warnings" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>True</cite> if one or more errors or warnings were found in the current section or its subsections.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.12)">bool</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="genophenocorr.model.html" class="btn btn-neutral float-left" title="genophenocorr.model package" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="genophenocorr.view.html" class="btn btn-neutral float-right" title="genophenocorr.view package" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>